File: Block Management/PacketFilter.m
//
//  PacketFilter.m
//  SelfControl
//
//  Created by Charles Stigler on 6/29/14.
//
//

#import "PacketFilter.h"

NSString* const kPfctlExecutablePath = @"/sbin/pfctl";
NSString* const kPFConfPath = @"/etc/pf.conf";
NSString* const kPFAnchorCommand = @"anchor \"org.eyebeam\"";

@implementation PacketFilter

NSFileHandle* appendFileHandle;

+ (BOOL)blockFoundInPF {
    // last try if we can't find a block anywhere: check the host file, and see if a block is in there
    NSString* pfConfContents = [NSString stringWithContentsOfFile: kPFConfPath encoding: NSUTF8StringEncoding error: NULL];
    if(pfConfContents != nil && [pfConfContents rangeOfString: kPFAnchorCommand].location != NSNotFound) {
        return YES;
    }

    return NO;
}

- (PacketFilter*)initAsAllowlist: (BOOL)allowlist {
	if (self = [super init]) {
		isAllowlist = allowlist;
		rules = [NSMutableString stringWithCapacity: 1000];
	}
	return self;
}

- (void)addBlockHeader:(NSMutableString*)configText {
	[configText appendString: @"# Options\n"
	 "set block-policy drop\n"
	 "set fingerprints \"/etc/pf.os\"\n"
	 "set ruleset-optimization basic\n"
	 "set skip on lo0\n"
	 "\n"
	 "#\n"
	 "# org.eyebeam ruleset for SelfControl blocks\n"
	 "#\n"];

	if (isAllowlist) {
		[configText appendString: @"block return out proto tcp from any to any\n"
		 "block return out proto udp from any to any\n"
		 "\n"];
	}
}
- (void)addAllowlistFooter:(NSMutableString*)configText {
	[configText appendString: @"pass out proto tcp from any to any port 53\n"];
	[configText appendString: @"pass out proto udp from any to any port 53\n"];
	[configText appendString: @"pass out proto udp from any to any port 123\n"];
	[configText appendString: @"pass out proto udp from any to any port 67\n"];
	[configText appendString: @"pass out proto tcp from any to any port 67\n"];
	[configText appendString: @"pass out proto udp from any to any port 68\n"];
	[configText appendString: @"pass out proto tcp from any to any port 68\n"];
	[configText appendString: @"pass out proto udp from any to any port 5353\n"];
	[configText appendString: @"pass out proto tcp from any to any port 5353\n"];
}

- (NSArray<NSString*>*)ruleStringsForIP:(NSString*)ip port:(NSInteger)port maskLen:(NSInteger)maskLen {
    NSMutableString* rule = [NSMutableString stringWithString: @"from any to "];

    if (ip) {
        [rule appendString: ip];
    } else {
        [rule appendString: @"any"];
    }

    if (maskLen) {
        [rule appendString: [NSString stringWithFormat: @"/%ld", (long)maskLen]];
    }

    if (port) {
        [rule appendString: [NSString stringWithFormat: @" port %ld", (long)port]];
    }

    if (isAllowlist) {
        return @[
            [NSString stringWithFormat: @"pass out proto tcp %@\n", rule],
            [NSString stringWithFormat: @"pass out proto udp %@\n", rule]
        ];
    } else {
        return @[
            [NSString stringWithFormat: @"block return out proto tcp %@\n", rule],
            [NSString stringWithFormat: @"block return out proto udp %@\n", rule]
        ];
    }
}
- (void)addRuleWithIP:(NSString*)ip port:(NSInteger)port maskLen:(NSInteger)maskLen {
    @synchronized(self) {
        NSArray<NSString*>* ruleStrings = [self ruleStringsForIP: ip port: port maskLen: maskLen];
        for (NSString* ruleString in ruleStrings) {
            if (appendFileHandle) {
                [appendFileHandle writeData: [ruleString dataUsingEncoding:NSUTF8StringEncoding]];
            } else {
                [rules appendString: ruleString];
            }
        }
    }
}

- (void)writeConfiguration {
	NSMutableString* filterConfiguration = [NSMutableString stringWithCapacity: 1000];

	[self addBlockHeader: filterConfiguration];
	[filterConfiguration appendString: rules];

	if (isAllowlist) {
		[self addAllowlistFooter: filterConfiguration];
	}

	[filterConfiguration writeToFile: @"/etc/pf.anchors/org.eyebeam" atomically: true encoding: NSUTF8StringEncoding error: nil];
}

- (void)enterAppendMode {
    if (isAllowlist) {
        NSLog(@"WARNING: Can't append rules to allowlist blocks - ignoring");
        return;
    }

    // open the file and prepare to write to the very bottom (no footer since it's not an allowlist)
    appendFileHandle = [NSFileHandle fileHandleForWritingAtPath: @"/etc/pf.anchors/org.eyebeam"];
    if (!appendFileHandle) {
        NSLog(@"ERROR: Failed to get handle for pf.anchors file while attempting to append rules");
        return;
    }

    [appendFileHandle seekToEndOfFile];
}
- (void)finishAppending {
    [appendFileHandle closeFile];
    appendFileHandle = nil;
}

- (void)appendRulesToCurrentBlockConfiguration:(NSArray<NSDictionary*>*)newEntryDicts {
    if (newEntryDicts.count < 1) return;
    if (isAllowlist) {
        NSLog(@"WARNING: Can't append rules to allowlist blocks - ignoring");
        return;
    }

    // open the file and prepare to write to the very bottom (no footer since it's not an allowlist)
    // NOTE FOR FUTURE: NSFileHandle can't append lines to the middle of the file anyway,
    // would need to read in the whole thing + write out again
    NSFileHandle* fileHandle = [NSFileHandle fileHandleForWritingAtPath: @"/etc/pf.anchors/org.eyebeam"];
    if (!fileHandle) {
        NSLog(@"ERROR: Failed to get handle for pf.anchors file while attempting to append rules");
        return;
    }

    [fileHandle seekToEndOfFile];
    for (NSDictionary* entryHostInfo in newEntryDicts) {
        NSString* hostName = entryHostInfo[@"hostName"];
        int portNum = [entryHostInfo[@"port"] intValue];
        int maskLen = [entryHostInfo[@"maskLen"] intValue];

        NSArray<NSString*>* ruleStrings = [self ruleStringsForIP: hostName port: portNum maskLen: maskLen];
        for (NSString* ruleString in ruleStrings) {
            [fileHandle writeData: [ruleString dataUsingEncoding:NSUTF8StringEncoding]];
        }
    }
    [fileHandle closeFile];
}

- (int)startBlock {
	[self addSelfControlConfig];
	[self writeConfiguration];

	NSArray* args = [@"-E -f /etc/pf.conf -F states" componentsSeparatedByString: @" "];

	NSTask* task = [[NSTask alloc] init];
	[task setLaunchPath: kPfctlExecutablePath];
	[task setArguments: args];

	NSPipe* inPipe = [[NSPipe alloc] init];
	NSFileHandle* readHandle = [inPipe fileHandleForReading];
	[task setStandardOutput: inPipe];
	[task setStandardError: inPipe];

	[task launch];
	NSString* pfctlOutput = [[NSString alloc] initWithData: [readHandle readDataToEndOfFile] encoding: NSUTF8StringEncoding];
	[readHandle closeFile];
	[task waitUntilExit];

	NSArray* lines = [pfctlOutput componentsSeparatedByString: @"\n"];
	for (NSString* line in lines) {
		if ([line hasPrefix: @"Token : "]) {
			[self writePFToken: [line substringFromIndex: [@"Token : " length]] error: nil];
			break;
		}
	}

	return [task terminationStatus];
}
- (int)refreshPFRules {
    NSArray* args = [@"-f /etc/pf.conf -F states" componentsSeparatedByString: @" "];

    NSTask* task = [[NSTask alloc] init];
    [task setLaunchPath: kPfctlExecutablePath];
    [task setArguments: args];
    [task launch];
    [task waitUntilExit];

    return [task terminationStatus];
}

- (void)writePFToken:(NSString*)token error:(NSError**)error {
	[token writeToFile: @"/etc/SelfControlPFToken" atomically: YES encoding: NSUTF8StringEncoding error: error];
}
- (NSString*)readPFToken:(NSError**)error {
	return [NSString stringWithContentsOfFile: @"/etc/SelfControlPFToken" encoding: NSUTF8StringEncoding error: error];
}

- (int)stopBlock:(BOOL)force {
	NSError* err;
	NSString* token = [self readPFToken: &err];

	[@"" writeToFile: @"/etc/pf.anchors/org.eyebeam" atomically: true encoding: NSUTF8StringEncoding error: nil];
	NSString* mainConf = [NSString stringWithContentsOfFile: @"/etc/pf.conf" encoding: NSUTF8StringEncoding error: nil];
	NSArray* lines = [mainConf componentsSeparatedByString: @"\n"];
	NSMutableString* newConf = [NSMutableString stringWithCapacity: [mainConf length]];
	for (NSString* line in lines) {
		if ([line rangeOfString: @"org.eyebeam"].location == NSNotFound) {
			[newConf appendFormat: @"%@\n", line];
		}
	}
	newConf = [[newConf stringByTrimmingCharactersInSet: [NSCharacterSet whitespaceAndNewlineCharacterSet]] mutableCopy];
	[newConf appendString: @"\n"];
	[newConf writeToFile: @"/etc/pf.conf" atomically: true encoding: NSUTF8StringEncoding error: nil];

	NSString* commandString;
	if ([token length] && !force) {
		commandString = [NSString stringWithFormat: @"-X %@ -f /etc/pf.conf", token];
	} else {
		commandString = @"-d -f /etc/pf.conf";
	}
	NSArray* args = [commandString componentsSeparatedByString: @" "];

	NSTask* task = [NSTask launchedTaskWithLaunchPath: kPfctlExecutablePath arguments: args];
	[task waitUntilExit];
	return [task terminationStatus];
}

- (void)addSelfControlConfig {
	NSMutableString* pfConf = [NSMutableString stringWithContentsOfFile: @"/etc/pf.conf" encoding: NSUTF8StringEncoding error: nil];

	if ([pfConf rangeOfString: @"/etc/pf.anchors/org.eyebeam"].location == NSNotFound) {
		[pfConf appendString: @"\n"
		 "anchor \"org.eyebeam\"\n"
		 "load anchor \"org.eyebeam\" from \"/etc/pf.anchors/org.eyebeam\"\n"];
	}

	[pfConf writeToFile: @"/etc/pf.conf" atomically: true encoding: NSUTF8StringEncoding error: nil];
}

- (BOOL)containsSelfControlBlock {
	NSString* mainConf = [NSString stringWithContentsOfFile: @"/etc/pf.conf" encoding: NSUTF8StringEncoding error: nil];
	return mainConf != nil && [mainConf rangeOfString: @"org.eyebeam"].location != NSNotFound;
}

@end

File: Common/Block/PacketFilter.swift
import Foundation

// MARK: - Protocol for testability

protocol PFConfigurationStore {
    var pfAnchorPath: String { get }
    var pfConfPath: String { get }
    var pfTokenPath: String { get }
}

struct DefaultPFConfigurationStore: PFConfigurationStore {
    var pfAnchorPath: String { "/etc/pf.anchors/\(StoneConstants.pfAnchorName)" }
    var pfConfPath: String { "/etc/pf.conf" }
    var pfTokenPath: String { "/etc/StonePFToken" }
}

// MARK: - PacketFilter

final class PacketFilter {

    private let isAllowlist: Bool
    private let store: PFConfigurationStore
    private var rules = ""
    private var appendFileHandle: FileHandle?
    private let lock = NSLock()

    private static let pfctlPath = "/sbin/pfctl"

    init(isAllowlist: Bool, store: PFConfigurationStore = DefaultPFConfigurationStore()) {
        self.isAllowlist = isAllowlist
        self.store = store
    }

    // MARK: - Static checks

    static func blockFoundInPF(store: PFConfigurationStore = DefaultPFConfigurationStore()) -> Bool {
        guard let contents = try? String(contentsOfFile: store.pfConfPath, encoding: .utf8) else {
            return false
        }
        return contents.contains("anchor \"\(StoneConstants.pfAnchorName)\"")
    }

    func containsStoneBlock() -> Bool {
        guard let contents = try? String(contentsOfFile: store.pfConfPath, encoding: .utf8) else {
            return false
        }
        return contents.contains(StoneConstants.pfAnchorName)
    }

    // MARK: - Rule generation

    private func ruleStrings(ip: String?, port: Int, maskLen: Int) -> [String] {
        var target = "from any to "
        target += ip ?? "any"
        if maskLen != 0 {
            target += "/\(maskLen)"
        }
        if port != 0 {
            target += " port \(port)"
        }

        if isAllowlist {
            return [
                "pass out proto tcp \(target)\n",
                "pass out proto udp \(target)\n"
            ]
        } else {
            return [
                "block return out proto tcp \(target)\n",
                "block return out proto udp \(target)\n"
            ]
        }
    }

    func addRule(ip: String?, port: Int, maskLen: Int) {
        lock.lock()
        defer { lock.unlock() }

        let strings = ruleStrings(ip: ip, port: port, maskLen: maskLen)
        for rule in strings {
            if let handle = appendFileHandle {
                if let data = rule.data(using: .utf8) {
                    handle.write(data)
                }
            } else {
                rules += rule
            }
        }
    }

    // MARK: - Configuration writing

    private func blockHeader() -> String {
        var header = """
        # Options
        set block-policy drop
        set fingerprints "/etc/pf.os"
        set ruleset-optimization basic
        set skip on lo0

        #
        # \(StoneConstants.pfAnchorName) ruleset for Stone blocks
        #\n
        """

        if isAllowlist {
            header += "block return out proto tcp from any to any\n"
            header += "block return out proto udp from any to any\n\n"
        }

        return header
    }

    private func allowlistFooter() -> String {
        return """
        pass out proto tcp from any to any port 53
        pass out proto udp from any to any port 53
        pass out proto udp from any to any port 123
        pass out proto udp from any to any port 67
        pass out proto tcp from any to any port 67
        pass out proto udp from any to any port 68
        pass out proto tcp from any to any port 68
        pass out proto udp from any to any port 5353
        pass out proto tcp from any to any port 5353\n
        """
    }

    func writeConfiguration() {
        var config = blockHeader()
        config += rules
        if isAllowlist {
            config += allowlistFooter()
        }
        try? config.write(toFile: store.pfAnchorPath, atomically: true, encoding: .utf8)
    }

    // MARK: - Append mode

    func enterAppendMode() {
        if isAllowlist {
            NSLog("WARNING: Can't append rules to allowlist blocks - ignoring")
            return
        }

        appendFileHandle = FileHandle(forWritingAtPath: store.pfAnchorPath)
        guard appendFileHandle != nil else {
            NSLog("ERROR: Failed to get handle for pf.anchors file while attempting to append rules")
            return
        }
        appendFileHandle?.seekToEndOfFile()
    }

    func finishAppending() {
        try? appendFileHandle?.close()
        appendFileHandle = nil
    }

    // MARK: - pf.conf management

    func addStoneConfig() {
        guard var pfConf = try? String(contentsOfFile: store.pfConfPath, encoding: .utf8) else {
            return
        }

        if !pfConf.contains(store.pfAnchorPath) {
            pfConf += "\n"
            pfConf += "anchor \"\(StoneConstants.pfAnchorName)\"\n"
            pfConf += "load anchor \"\(StoneConstants.pfAnchorName)\" from \"\(store.pfAnchorPath)\"\n"
        }

        try? pfConf.write(toFile: store.pfConfPath, atomically: true, encoding: .utf8)
    }

    // MARK: - Start / Stop

    @discardableResult
    func startBlock() -> Int32 {
        addStoneConfig()
        writeConfiguration()

        let args = ["-E", "-f", store.pfConfPath, "-F", "states"]
        let (status, output) = runPfctl(args)

        // Parse and save the token
        let lines = output.components(separatedBy: "\n")
        for line in lines {
            if line.hasPrefix("Token : ") {
                let token = String(line.dropFirst("Token : ".count))
                writePFToken(token)
                break
            }
        }

        return status
    }

    @discardableResult
    func refreshPFRules() -> Int32 {
        let args = ["-f", store.pfConfPath, "-F", "states"]
        let (status, _) = runPfctl(args)
        return status
    }

    @discardableResult
    func stopBlock(force: Bool) -> Int32 {
        let token = readPFToken()

        // Clear anchor file
        try? "".write(toFile: store.pfAnchorPath, atomically: true, encoding: .utf8)

        // Remove Stone lines from pf.conf
        if let mainConf = try? String(contentsOfFile: store.pfConfPath, encoding: .utf8) {
            let lines = mainConf.components(separatedBy: "\n")
            var newConf = lines
                .filter { !$0.contains(StoneConstants.pfAnchorName) }
                .joined(separator: "\n")
            newConf = newConf.trimmingCharacters(in: .whitespacesAndNewlines) + "\n"
            try? newConf.write(toFile: store.pfConfPath, atomically: true, encoding: .utf8)
        }

        let args: [String]
        if let token = token, !token.isEmpty, !force {
            args = ["-X", token, "-f", store.pfConfPath]
        } else {
            args = ["-d", "-f", store.pfConfPath]
        }

        let (status, _) = runPfctl(args)
        return status
    }

    // MARK: - Token persistence

    private func writePFToken(_ token: String) {
        try? token.write(toFile: store.pfTokenPath, atomically: true, encoding: .utf8)
    }

    private func readPFToken() -> String? {
        return try? String(contentsOfFile: store.pfTokenPath, encoding: .utf8)
    }

    // MARK: - Process helper

    private func runPfctl(_ arguments: [String]) -> (Int32, String) {
        let task = Process()
        task.executableURL = URL(fileURLWithPath: PacketFilter.pfctlPath)
        task.arguments = arguments

        let pipe = Pipe()
        task.standardOutput = pipe
        task.standardError = pipe

        do {
            try task.run()
        } catch {
            NSLog("ERROR: Failed to launch pfctl: %@", error.localizedDescription)
            return (-1, "")
        }

        let data = pipe.fileHandleForReading.readDataToEndOfFile()
        task.waitUntilExit()
        let output = String(data: data, encoding: .utf8) ?? ""
        return (task.terminationStatus, output)
    }
}
